### ISSUE TITLE
Autoloading in development is not thread-safe

### ISSUE THREAD

**Issue by tmtrademarked** [bigcommerce/gruf#174]
## Please describe the issue

When running a grpc server with more than one thread locally, multiple simultaneous requests can come in at once. In this case, each request will trigger a call to `Gruf.autoloaders.reload`, which attempts to reload the Zeitwerk environment. The result is that while processing requests, some calls will inconsistently fail with Ruby class loading errors.

## How to Reproduce

It was hard for me to reproduce this reliably. Eventually what I did was to create a test script that would invoke my GRPC endpoint, and ran it on a tight loop:
```
for i in {1..10}; do ruby tools/grpc_test_cart.rb 2411031 &; done
```

This results in *some* successful requests, and some failures with errors like the following:
```
.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/grpc-1.50.0/src/ruby/lib/grpc/generic/active_call.rb:29:in `check_status': 2:NameError: uninitialized constant CartServiceRpcController (GRPC::Unknown)

      Object.const_get(camel_cased_word)
            ^^^^^^^^^^. debug_error_string:{UNKNOWN:Error received from peer ipv6:%5B::1%5D:50051 {grpc_message:"NameError: uninitialized constant CartServiceRpcController\\n\\n      Object.const_get(camel_cased_word)\\n            ^^^^^^^^^^", grpc_status:2, created_time:"2022-11-14T15:18:44.571699-05:00"}}
	from /Users/tom.wilson/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/grpc-1.50.0/src/ruby/lib/grpc/generic/active_call.rb:186:in `attach_status_results_and_complete_call'
	from /Users/tom.wilson/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/grpc-1.50.0/src/ruby/lib/grpc/generic/active_call.rb:382:in `request_response'
	from /Users/tom.wilson/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/grpc-1.50.0/src/ruby/lib/grpc/generic/client_stub.rb:180:in `block in request_response'
	from /Users/tom.wilson/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/grpc-1.50.0/src/ruby/lib/grpc/generic/interceptors.rb:170:in `intercept!'
	from /Users/tom.wilson/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/grpc-1.50.0/src/ruby/lib/grpc/generic/client_stub.rb:179:in `request_response'
	from /Users/tom.wilson/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/grpc-1.50.0/src/ruby/lib/grpc/generic/service.rb:171:in `block (3 levels) in rpc_stub_class'
	from tools/grpc_get_most_recent_cart.rb:20:in `main'
	from tools/grpc_get_most_recent_cart.rb:24:in `<main>'
```

To work around this, I tested a few solutions. Simply removing the call to `Gruf.autoloaders.reload` from the base controller's `#call` method fixes this, at the cost of disallowing autoload. This would be an acceptable workaround, but it's difficult to reach the `Gruf::Autoloader` instantiation hook. Is there a way to configure this better?

## What _should_ happen?

All the requests should successfully complete.

## Anything else we should know?

* gRPC version - 1.50.0
* Ruby version - 3.1.2

**Comment by tmtrademarked** [comment]
Hmm - unfortunately, it doesn't seem to fix it for me. After I unwind my local patches and update to 2.16.2, I'm still seeing some of my 10 threads hit class loading issues. Specifically, it looks like they're all getting name errors when attempting to resolve the RpcController class, which smells like the same Zeitwerk issue somehow:

```
/Users/tom.wilson/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/grpc-1.50.0/src/ruby/lib/grpc/generic/active_call.rb:29:in `check_status': 2:NameError: uninitialized constant CartServiceRpcController (GRPC::Unknown)

      Object.const_get(camel_cased_word)
            ^^^^^^^^^^. debug_error_string:{UNKNOWN:Error received from peer ipv6:%5B::1%5D:50051 {grpc_message:"NameError: uninitialized constant CartServiceRpcController\\n\\n      Object.const_get(camel_cased_word)\\n            ^^^^^^^^^^", grpc_status:2, created_time:"2022-11-22T11:29:00.900444-05:00"}}
	from /Users/tom.wilson/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/grpc-1.50.0/src/ruby/lib/grpc/generic/active_call.rb:186:in `attach_status_results_and_complete_call'
	from /Users/tom.wilson/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/grpc-1.50.0/src/ruby/lib/grpc/generic/active_call.rb:382:in `request_response'
	from /Users/tom.wilson/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/grpc-1.50.0/src/ruby/lib/grpc/generic/client_stub.rb:180:in `block in request_response'
	from /Users/tom.wilson/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/grpc-1.50.0/src/ruby/lib/grpc/generic/interceptors.rb:170:in `intercept!'
	from /Users/tom.wilson/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/grpc-1.50.0/src/ruby/lib/grpc/generic/client_stub.rb:179:in `request_response'
	from /Users/tom.wilson/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/grpc-1.50.0/src/ruby/lib/grpc/generic/service.rb:171:in `block (3 levels) in rpc_stub_class'
	from tools/grpc_get_most_recent_cart.rb:20:in `main'
	from tools/grpc_get_most_recent_cart.rb:24:in `<main>'
```

I can't logically explain why that fix doesn't work for me, though. It *seems* like it should, but maybe there's something else, either in our local setup or in Gruf?

**Comment by leonid-shevtsov** [comment]
I encountered the same issue. To repro the errors:

- boot server in dev mode
- introduce some changes (to trigger reloading)
- have the client make 2 or more requests in parallel (this is the crucial step)
- observe the missing constant error describe above

To resolve, you need to make code reloading happen _before_ the constant is resolved, and certainly not _during_.

Fixed in #178
